tests/generic_config_updater/test_syslog.py

```diff
@@ -2,66 +2,136 @@
+
+@pytest.fixture(scope="module", params=["config_cleanup", "config_add_default"])
+def init_syslog_config(request):
+    return request.param
+
+def syslog_config_cleanup(duthost, cfg_facts):
+    """ Cleanup syslog server
+
+    Sample output
+    admin@vlab-01:~$ show runningconfiguration syslog
+    Syslog Servers
+    ----------------
+    """
+    syslog_servers = cfg_facts.get('SYSLOG_SERVER', {})
+    for syslog_server in syslog_servers:
+        del_syslog_server = duthost.shell("sudo config syslog del {}".format(syslog_server),
+            module_ignore_errors=True)
+        pytest_assert(not del_syslog_server['rc'],
+            "syslog server '{}' is not deleted successfully".format(syslog_server))
+
+
+def syslog_config_add_default(duthost):
+    """ Add default v4 and v6 syslog server to config
+
+    Sample output
+    admin@vlab-01:~$ show runningconfiguration syslog
+    Syslog Servers
+    ----------------
+    [10.0.0.5]
+    [cc98:2008::1]
+    """
+    for syslog_server in [SYSLOG_DUMMY_IPV4_SERVER, SYSLOG_DUMMY_IPV6_SERVER]:
+        add_syslog_server = duthost.shell("sudo config syslog add {}".format(syslog_server),
+            module_ignore_errors=True)
+        pytest_assert(not add_syslog_server['rc'],
+            "syslog server '{}' is not deleted successfully".format(syslog_server))
 
 @pytest.fixture(scope="module")
-def setup_env(duthosts, rand_one_dut_hostname, cfg_facts):
+def original_syslog_servers(duthost, tbinfo):
+    """A module level fixture to store original syslog servers info
+    """
+    mg_facts = duthost.get_extended_minigraph_facts(tbinfo)
+
+    original_syslog_servers = []
+    for syslog_server in mg_facts['syslog_servers']:
+        original_syslog_servers.append(syslog_server)
+
+    return original_syslog_servers
+
+def get_current_syslog_servers(duthost):
+    """Get current syslog servers from running host
+    """
+    cmds = "show runningconfiguration syslog"
+    output = duthost.shell(cmds)
+
+    pytest_assert(not output['rc'],
+        "'{}' is not running successfully".format(cmds)
+    )
+
+    # Jump over introductory printout
+    output_lines = output['stdout'].splitlines()[2:]
+    current_syslog_servers = []
+    for line in output_lines:
+        # remove enclosed sqaure bracket
+        current_syslog_servers.append(line[1:-1])
+    return current_syslog_servers
+
+@pytest.fixture(autouse=True)
+def setup_env(duthosts, rand_one_dut_hostname, cfg_facts, init_syslog_config, original_syslog_servers):
     """
     Setup/teardown fixture for syslog config
     Args:
         duthosts: list of DUTs.
         rand_selected_dut: The fixture returns a randomly selected DuT.
         cfg_facts: config facts for selected DUT
+        init_syslog_config: set up the initial syslog config for test
+        original_syslog_servers: original syslog servers stored in config
     """
     duthost = duthosts[rand_one_dut_hostname]
+    create_checkpoint(duthost)
```

Comment:
>create_checkpoint [](http://example.com/codeflow?start=4&length=17)

Since the apply patch test starting from `yield`, should we create checkpoint just before `yield`?

Reply:
We want it to rolled back to original setup. Cuz the test env is based on original setup.
If we created before yield, it will rollback to `config_cleanup` or `config_add_default` and we will need to add rollback check for each setup.
To move the checkpoint just before yield, we can either 
1. move each setup code inside test, 
2. or do another rollback to original setup.
or we can keep the current setup.

Reply:
It's a good idea to rollback twice.

Reply:
Add second rollback